calico_containers/tests/st/utils/workload.py

```diff
@@ -125,15 +128,44 @@ def assert_can_ping(self, ip, retries=0):
         command = ' '.join(args)
 
         ping = partial(self.execute, command)
-        return retry_until_success(ping,
+        return ping
+
+    def assert_can_ping(self, ip, retries=0):
+        """
+        Execute a ping from this workload to the ip. Assert than a workload
+        can ping an IP. Use retries to compensate for network uncertainty and
+        convergence.
+
+        :param ip:  The IP address (str or IPAddress) to ping.
+        :param retries: The number of retries.
+        :return: None.
+        """
+        return retry_until_success(self._get_ping_function(ip),
                                    retries=retries,
-                                   ex_class=CalledProcessError)
+                                   ex_class=CommandExecError)
 
     def assert_cant_ping(self, ip, retries=0):
-        for retry in range(retries + 1):
-            self.assert_can_ping(ip)
-            if retry >= retries:
-                raise Exception("Workload can unexpectedly ping %s" % ip)
+        """
+        Execute a ping from this workload to the ip.  Assert that the workload
+        cannot ping an IP.  Use retries to compensate for network uncertainty
+        and convergence.
+
```

Comment:
This function is odd.  The goal is to verify that containers cannot reach each other.  Network uncertainty results in dropped packets, so retrying the function will just make it more likely that a dropped packet will represent a false positive.

If we're going to support retries at all on this function (and maybe we shouldn't), then you need separate numbers for

 1. Number of successful pings before giving up and failing.
 2. Number of dropped pings before declaring a pass.

The first gives you control over how long to wait for network convergence, the second allows you to protect yourself from random network events.  Note that for `assert_can_ping` these factors act in concert, here they act opposite directions.